File: tests/mathematical/matrix_multiply/matrix_multiply.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func createMatrix(rows, cols int) [][]float64 {
	matrix := make([][]float64, rows)
	for i := range matrix {
		matrix[i] = make([]float64, cols)
		for j := range matrix[i] {
			matrix[i][j] = rand.Float64() * 100
		}
	}
	return matrix
}

func multiplyMatrices(a, b [][]float64) [][]float64 {
	rowsA := len(a)
	colsA := len(a[0])
	colsB := len(b[0])
	
	result := make([][]float64, rowsA)
	for i := range result {
		result[i] = make([]float64, colsB)
	}
	
	for i := 0; i < rowsA; i++ {
		for j := 0; j < colsB; j++ {
			for k := 0; k < colsA; k++ {
				result[i][j] += a[i][k] * b[k][j]
			}
		}
	}
	
	return result
}

func main() {
	rand.Seed(time.Now().UnixNano())
	size := 200 // Matrix size (200x200)
	
	fmt.Printf("Multiplying two %dx%d matrices...\n", size, size)
	
	// Create matrices
	createStart := time.Now()
	matrixA := createMatrix(size, size)
	matrixB := createMatrix(size, size)
	createTime := time.Since(createStart)
	
	// Multiply matrices
	multiplyStart := time.Now()
	result := multiplyMatrices(matrixA, matrixB)
	multiplyTime := time.Since(multiplyStart)
	
	totalTime := createTime + multiplyTime
	
	// Verify result dimensions
	resultRows := len(result)
	resultCols := len(result[0])
	
	fmt.Printf("Result: %dx%d matrix\n", resultRows, resultCols)
	fmt.Printf("Sample result[0][0]: %.6f\n", result[0][0])
	fmt.Println("Timing:")
	fmt.Printf("  Matrix creation: %.6f seconds\n", createTime.Seconds())
	fmt.Printf("  Matrix multiplication: %.6f seconds\n", multiplyTime.Seconds())
	fmt.Printf("  Total time: %.6f seconds\n", totalTime.Seconds())
}

File: tests/mathematical/matrix_multiply/matrix_multiply.rs
use std::time::Instant;
use rand::{thread_rng, Rng};

fn create_matrix(rows: usize, cols: usize) -> Vec<Vec<f64>> {
    let mut rng = thread_rng();
    (0..rows)
        .map(|_| {
            (0..cols)
                .map(|_| rng.gen_range(0.0..100.0))
                .collect()
        })
        .collect()
}

fn multiply_matrices(a: &[Vec<f64>], b: &[Vec<f64>]) -> Vec<Vec<f64>> {
    let rows_a = a.len();
    let cols_a = a[0].len();
    let cols_b = b[0].len();
    
    let mut result = vec![vec![0.0; cols_b]; rows_a];
    
    for i in 0..rows_a {
        for j in 0..cols_b {
            for k in 0..cols_a {
                result[i][j] += a[i][k] * b[k][j];
            }
        }
    }
    
    result
}

fn main() {
    let size = 200;  // Matrix size (200x200)
    
    println!("Multiplying two {}x{} matrices...", size, size);
    
    // Create matrices
    let create_start = Instant::now();
    let matrix_a = create_matrix(size, size);
    let matrix_b = create_matrix(size, size);
    let create_time = create_start.elapsed();
    
    // Multiply matrices
    let multiply_start = Instant::now();
    let result = multiply_matrices(&matrix_a, &matrix_b);
    let multiply_time = multiply_start.elapsed();
    
    let total_time = create_time + multiply_time;
    
    // Verify result dimensions
    let result_rows = result.len();
    let result_cols = result[0].len();
    
    println!("Result: {}x{} matrix", result_rows, result_cols);
    println!("Sample result[0][0]: {:.6}", result[0][0]);
    println!("Timing:");
    println!("  Matrix creation: {:.6} seconds", create_time.as_secs_f64());
    println!("  Matrix multiplication: {:.6} seconds", multiply_time.as_secs_f64());
    println!("  Total time: {:.6} seconds", total_time.as_secs_f64());
}